### PR TITLE
Terrain

### PR DESCRIPTION
**Pull request by super626**

### CODE REVIEW

File: cocos/3d/CCTerrain.cpp [hunk 1]
```diff
@@ -0,0 +1,1380 @@
+#include "CCTerrain.h"
```
Comment (minggo) [inline cocos/3d/CCTerrain.cpp:25]: Please add license information.

File: cocos/3d/CCTerrain.h [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    *TerrainData
+    *This TerrainData struct warp all parameter that Terrain need to create
+    */
+    struct CC_DLL TerrainData
+    {
+        /*Constructors*/
+        TerrainData();
+        TerrainData(const char* heightMapsrc, const char * textureSrc, const Size & chunksize = Size(32,32), float mapHeight = 2, float mapScale = 0.1);
+        TerrainData(const char* heightMapsrc, const char * alphamap, const DetailMap& detail1,const DetailMap& detail2, const DetailMap& detail3, const DetailMap& detail4, const Size & chunksize = Size(32,32), float mapHeight = 2, float mapScale = 0.1);
+        TerrainData(const char* heightMapsrc, const char * alphamap, const DetailMap& detail1,const DetailMap& detail2, const DetailMap& detail3, const Size & chunksize = Size(32,32), float mapHeight = 2, float mapScale = 0.1);
```
Comment (minggo) [inline cocos/3d/CCTerrain.h:121]: Should add comment for every method, or it will not be shown in document.

File: cocos/3d/CCTerrain.cpp [hunk 3]
```diff
@@ -0,0 +1,1380 @@
...
+        }
+    }
+}
+
+void Terrain::QuadTree::preCalculateAABB(const Mat4 & worldTransform)
+{
+
+        _worldSpaceAABB = _aabb;
+        _worldSpaceAABB.transform(worldTransform);
+    if(!_isTerminal){
```
Comment (minggo) [inline cocos/3d/CCTerrain.cpp:1367]: Wrong indent.

File: cocos/3d/CCTerrain.h [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+public:
+    /*init function*/
+    bool init();
+    void initHeightMap(const char* heightMap);
+    /*create entry*/
+    static Terrain * create(TerrainData &parameter, CrackFixedType fixedType = CrackFixedType::INCREASE_LOWER);
+    /*get specified position's height mapping to the terrain*/
+    float getHeight(float x, float z, Vec3 * normal= nullptr);
+    float getHeight(Vec2 pos, Vec3*Normal = nullptr);
+    Vec3 getNormal(int pixel_x, int pixel_y);
```
Comment (minggo) [inline cocos/3d/CCTerrain.h]: pixel_x -> pixelX

File: cocos/3d/CCTerrain.h [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+        void calculateSlope();
+        /*current LOD of the chunk*/
+        int _currentLod;
+
+        int _oldLod;
+
+        int _neighborOldLOD[4];
+        /*the left,right,front,back neighbors*/
+        Chunk * left;
```
Comment (minggo) [inline cocos/3d/CCTerrain.h]: left -> _left

File: cocos/3d/CCTerrain.h [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+
+        //the position
+        int pos_x;
+        int pos_y;
+        /*parent terrain*/
+        Terrain * _terrain;
+        /*chunk size*/
+        Size _size;
+        /*chunk's estimated slope*/
+        float slope;
```
Comment (minggo) [inline cocos/3d/CCTerrain.h]: slope -> _slope

**Comment by minggo** [comment]
And please compress these commits to one commit.

**Comment by super626** [comment]
@minggo 
Ok, Thanks.

File: cocos/3d/CCTerrain.cpp (hunk 1 above)
Comment (tangziwen) [inline cocos/3d/CCTerrain.cpp:25]: already add

File: cocos/3d/CCTerrain.h [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+        Chunk();
+        /**destructor*/
+        ~Chunk();
+        /*vertices*/
+        std::vector<TerrainVertexData> vertices;
+        /*LOD indices*/
+        struct LOD{
+            std::vector<GLushort> indices;
+        };
+        GLuint vbo[2];
```
Comment (minggo) [inline cocos/3d/CCTerrain.h:195]: vbo -> _ vbo

File: cocos/3d/CCTerrain.h [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+        /*the left,right,front,back neighbors*/
+        Chunk * _left;
+        Chunk * _right;
+        Chunk * _front;
+        Chunk * _back;
+
+        QuadTree * _parent;
+
+        /**the position X in terrain space*/
+        int pos_x;
```
Comment (minggo) [inline cocos/3d/CCTerrain.h:233]: pos_x -> _posX
Please modify all the members.

File: cocos/3d/CCTerrain.h [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+     */
+    void setSkirtHeightRatio(float ratio);
+
+    /**
+     * get the terrain's quad tree which is also the root node.
+     */
+    QuadTree * getQuadTree();
+
+    //following methods are internal use only
+    ChunkIndices lookForIndicesLODSkrit(int selfLod, bool * result);
```
Comment (samuele3hu) [inline cocos/3d/CCTerrain.h:385]: These methods are internal used only,but the visibleness of these methods are public.

**Pull request merged**